### DISCUSSION TITLE
brainstorm idea list - a place for new ideas

### DISCUSSION THREAD

**Discussion by dc-redeclipse** [redeclipse/base#1242] [Ideas and Feature Requests]
I accidentally brainstormed some ideas last night on misreading the utility discussion... but the ideas maybe interesting for you to think a bit more about it.

- different turret models depending on the weapon type:
 - gun turret model only on type smg, pistol, rifle, shotgun 
 - rocket launcher turret model only on weapon type rocket
 - flamer turret model only on weapon type flamer
- possibility to back flip and /180 turn on while wall climb
- triggered rails
- WebRTC for in-game and out- to in-game communications connections (to replace irc)
- NPCs (kill able pedestrians idea)
- single and multiplayer missions
- matchmaking system / kinda player lobby
- server site executable scripts in maps (maybe an flag on which instance the script will be executed)
- python <-> cube script wrapper
- ranking system to motivate player for playing RE more often maybe in combinations with achievements like cool vanities only with specific ranks
- server flag to store costume maps uploaded by players and make them select able
- new game mode mutant: 30% player vs 70% players only with claws slower less life points catch the player turn them to mutant and win after mutating all players, players win only after surviving the hole match at least only one
- gravity material, material with other configurable gravity perfect for parkour
- slow motion power item for parkour
- power item for parkour
- knife as an alternative for kick for near fights but on same player position level

**Comment by dc-redeclipse** [comment]
RaZgRiZ had a very good Idea how to prevent var over-writings to the client. We should implement a method that detects a change set a flag and will clean up on every disconnect if the flag is active.

**Reply by janitor-raus** [reply]
The premise, at least in paper, is to keep track of the last server which enforced custom vars and such upon a connecting client, and preserve the state of the client's vars at the time. When the client is disconnected from that server, a call will be made to clear the custom vars by restoring the originals. Similarly, the client should also perform this kind of cleanup check during its boot to ensure that there's no leftovers in place in the event of a crash that would have skipped the on-disconnect call.

**Reply by dc-redeclipse** [reply]
yeah we really need that soon! :D thank you for that!

**Comment by janitor-raus** [comment]
>     * different turret models depending on the weapon type:
>     * gun turret model only on type smg, pistol, rifle, shotgun
>     * rocket launcher turret model only on weapon type rocket
>     * flamer turret model only on weapon type flamer

I love turrets, gibbe plx

>     * possibility to back flip and /180 turn on while wall climb

The purpose being? Aside from making cool cinematics, I'm not sure there's a lot of gameplay value to doing wallrun backflips.

>     * triggered rails

I think that's planned already, just a question of when and how to tackle it.

>     * WebRTC for in-game and out- to in-game communications connections (to replace irc)

That could work if one could hook a discord bot to a server as a relay.

>     * NPCs (kill able pedestrians idea)

Sure, but A) I'm not sure if their animation even syncs up with other players or if its entirely client-side (probably the latter), and B) supposing you fix A, what is the point of murdering NPC robots?

>     * single and multiplayer missions

What's their purpose though.

>     * matchmaking system / kinda player lobby

Not really a concern when your playerbase is still tiny.

>     * server site executable scripts in maps (maybe an flag on which instance the script will be executed)

If triggers can be made to run scripts and functions are put in place to call upon the server to apply an action to a particular player or some sort of effect, it should be doable. Requiring this to become an entity or be attached to one isn't realistic I think.

>     * python <-> cube script wrapper

Handling cubescript is by itself difficult as it is, making some sort of intepreter for python as well will complicate things even further, especially if you expect the two to interact with each other. I take it this falls in line with your idea of server mods?

>     * ranking system to motivate player for playing RE more often maybe in combinations with achievements like cool vanities only with specific ranks

That might well have to wait until RE expands its playerbase further, having ranking matches between all of 20 people doesn't exactly seem all that riveting, especially with a fluctuating gameplay model.

>     * server flag to store costume maps uploaded by players and make them select able

* custom, and the idea isn't bad to maintain a cache, if you can figure out the specifics of how it'd work. It doesn't seem like something that'd be too high on a priority list, more like a nice-to-have.

>     * new game mode mutant: 30% player vs 70% players only with claws slower less life points catch the player turn them to mutant and win after mutating all players, players win only after surviving the hole match at least only one

Zombie modes are overrated, and I don't care if you call them mutants :P 

>     * gravity material, material with other configurable gravity perfect for parkour

Could potentially be a trigger event that can apply to any one player, or multiple/all of them. IIRC gravity is already a map var or some such, you could just implement a more granular control. Having an extra mat might be difficult. What color would you give it?

>     * slow motion power item for parkour

That will not work in multi-player..

>     * power item for parkour

And it does what exactly? Was it just a suggestion for some sort of powerup, ideas welcome, pitch in, pick the popular one?

>     * knife as an alternative for kick for near fights but on same player position level

Sorry what?

**Comment by ghost** [comment]
Ability to create single/multi-player missions may attract more people to the game. 
Personally I do not like Red Eclipse gameplay (sorry to say that) but I like editor and graphics. I would like to create some levels inspired by DeusEx with some simple scripts but cubescript freaking me out :)
I would love to stick to the game as a creator if I see that it is much simpler and fun than taking some other engine. 

Introduction of some alternative(but more famous) scripting language may also help (lua, python or javascript) or at least a transpiler for their subsets.

**Reply by dc-redeclipse** [reply]
thanks for your replay and interest into the project.
there are a lot of idea to make missions like that. Here an old [example](https://www.youtube.com/watch?v=TovHiacgp7I).
 
There is a [plot](https://github.com/redeclipse/base/discussions/1317) now written down for a story to base missions on and simular stuf like that. There are many ideas out (some of them in my head also) how missions can look like and how can cube script make those interactiv with typical missions atmospherics.
Sure we need dialogs, sounds, some misson UI  stuff and the storyline need some extentions but these idea are long enough in my head for that.

With the new editui making maps was never been simplier then it is to day. for sure to fit the game style and estatics you still need to do a lot of work to match other official maps in the game but many things are more easy done then before.>

Introduction of something new is one thing, the idea of an working python <-> cube script wrapper for example should help to get around makeing things all over new. CS is hard sometimes but also things work fine if you learn how to deal with it and bring some room for pain  while stumbiling about none conevtioninal habits of CS.
